pkg/rpcclient/actor/waiter.go

```diff
@@ -0,0 +1,297 @@
+
+// newWaiter creates Waiter instance. It can be either websocket-based or
+// polling-base, otherwise Waiter stub is returned.
+func newWaiter(ra RPCActor, v *result.Version) Waiter {
+	if eventW, ok := ra.(RPCEventWaiter); ok {
+		return &EventWaiter{
+			ws: eventW,
+			polling: &PollingWaiter{
+				polling: eventW,
+				version: v,
+			},
+		}
+	}
+	if pollW, ok := ra.(RPCPollingWaiter); ok {
+		return &PollingWaiter{
+			polling: pollW,
+			version: v,
+		}
+	}
+	return NewNullWaiter()
+}
+
+// NewNullWaiter creates an instance of Waiter stub.
+func NewNullWaiter() NullWaiter {
+	return NullWaiter{}
+}
+
+// Wait implements Waiter interface.
+func (NullWaiter) Wait(h util.Uint256, vub uint32, err error) (*state.AppExecResult, error) {
+	return nil, ErrAwaitingNotSupported
+}
+
+// WaitAny implements Waiter interface.
+func (NullWaiter) WaitAny(ctx context.Context, vub uint32, hashes ...util.Uint256) (*state.AppExecResult, error) {
+	return nil, ErrAwaitingNotSupported
+}
+
+// NewPollingWaiter creates an instance of Waiter supporting poll-based transaction awaiting.
+func NewPollingWaiter(waiter RPCPollingWaiter) (*PollingWaiter, error) {
+	v, err := waiter.GetVersion()
+	if err != nil {
+		return nil, err
+	}
+	return &PollingWaiter{
+		polling: waiter,
+		version: v,
+	}, nil
+}
+
+// Wait implements Waiter interface.
+func (w *PollingWaiter) Wait(h util.Uint256, vub uint32, err error) (*state.AppExecResult, error) {
+	if err != nil {
+		return nil, err
+	}
+	return w.WaitAny(context.TODO(), vub, h)
+}
+
+// WaitAny implements Waiter interface.
+func (w *PollingWaiter) WaitAny(ctx context.Context, vub uint32, hashes ...util.Uint256) (*state.AppExecResult, error) {
+	var (
+		currentHeight uint32
+		failedAttempt int
+		pollTime      = time.Millisecond * time.Duration(w.version.Protocol.MillisecondsPerBlock) / 2
+	)
+	if pollTime == 0 {
+		pollTime = time.Second
+	}
+	timer := time.NewTicker(pollTime)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			blockCount, err := w.polling.GetBlockCount()
+			if err != nil {
+				failedAttempt++
+				if failedAttempt > PollingWaiterRetryCount {
+					return nil, fmt.Errorf("failed to retrieve block count: %w", err)
+				}
+				continue
+			}
+			failedAttempt = 0
+			if blockCount-1 > currentHeight {
+				currentHeight = blockCount - 1
+			}
+			t := trigger.Application
+			for _, h := range hashes {
+				res, err := w.polling.GetApplicationLog(h, &t)
+				if err == nil {
+					return &state.AppExecResult{
+						Container: res.Container,
+						Execution: res.Executions[0],
+					}, nil
+				}
+			}
+			if currentHeight >= vub {
+				return nil, ErrTxNotAccepted
+			}
+		case <-w.polling.Context().Done():
+			return nil, fmt.Errorf("%w: %v", ErrContextDone, w.polling.Context().Err())
+		case <-ctx.Done():
+			return nil, fmt.Errorf("%w: %v", ErrContextDone, ctx.Err())
+		}
+	}
+}
+
+// NewEventWaiter creates an instance of Waiter supporting websocket event-based transaction awaiting.
```

Comment:
We should mention that it falls back to polling when subscription fails.

Reply:
Fixed.